pkg/provider/core.go

```diff
@@ -398,6 +400,32 @@ func decodeProviderSpec(machineClass *v1alpha1.MachineClass) (*api.EquinixMetalP
 	return providerSpec, nil
 }
 
+func createDeviceWithReservations(svc packngo.DeviceService, createRequest *packngo.DeviceCreateRequest, reservationIDs []string, reservedOnly bool) (device *packngo.Device, err error) {
+	// if there were no reservation IDs and I didn't ask for reservedOnly, then just create one on-demand and return
+	if len(reservationIDs) == 0 && !reservedOnly {
+		device, _, err = svc.Create(createRequest)
+		return device, err
+	}
+
+	// if we got here, we either had some reservation IDs, or we were asked to do reserved only.
+	// In both cases, we try reservations first.
+	for _, resID := range reservationIDs {
+		createRequest.HardwareReservationID = resID
+		device, _, err = svc.Create(createRequest)
+		// if no error, we got the device, return it
+		if err == nil {
+			return device, err
+		}
```

Comment:
Is there a way to handle more specifically the error we are expecting ("reservation ID already in use" or whatever is returned by the API)?

Reply:
There is, same as the [other comment](https://github.com/gardener/machine-controller-manager-provider-equinix-metal/pull/7#discussion_r695467840). The proposed implementation in this PR works, somewhat inefficiently, and without precise enough errors. I prefer to get this in as working, then we can optimize it in small, lightweight PRs.

Reply:
OK